proposals/0000-overloaded-record-fields.rst

```diff
@@ -0,0 +1,457 @@
+
+
+Recap: OverloadedLabels
+~~~~~~~~~~~~~~~~~~~~~~~
+
+The ``OverloadedLabels`` extension, also introduced in GHC 8.0.1,
+provides a way to indicate that an identifier should be resolved using
+type inference.  This extension provides a new syntax (using a ``#``
+prefix) for "labels", identifiers whose meaning depends on their type
+as well as their name.  A label ``#foo`` has type ``IsLabel "foo" t =>
+t`` where ``IsLabel`` is an ordinary type class (rather like
+``IsString`` but with an additional type parameter for the text of the
+label).
+
+This extension has already been used by
+various anonymous records libraries, such as
+`bookkeeper <https://hackage.haskell.org/package/bookkeeper>`_,
+`rawr <http://hackage.haskell.org/package/rawr>`_,
+`ruin <http://hackage.haskell.org/package/ruin>`_ and no doubt others.
+However, it is not so easy to use with normal Haskell record
+datatypes, because there is no built-in support for polymorphism over
+record selectors.
+
+We want to give an instance
+
+.. code-block:: haskell
+
+    instance ... => IsLabel x (r -> a)
+
+to make it possible to interpret a label (e.g. ``#personId``) as a
+record selector function, determining the record and field types as
+part of the type inference process.  For example, the expression
+
+.. code-block:: haskell
+
+  \ (p :: Person) -> #personId p
+
+will be accepted with inferred type ``Person -> Int``.  In addition,
+the proposed change enables definitions that are polymorphic over
+record fields.
+
+But what should go in the instance context? In the following section,
+we introduce a new built-in class ``HasField`` for this purpose.
+
+
+Proposed Change
+---------------
+
+HasField class
+~~~~~~~~~~~~~~
+
+The new module ``GHC.Records`` defines the following:
+
+.. code-block:: haskell
+
+  class HasField (x :: k) r a | x r -> a where
+    getField :: r -> a
+
+A ``HasField x r a`` constraint represents the fact that ``x`` is a
+field of type ``a`` belonging to a record type ``r``.  The
+``getField`` method gives the record selector function.
+
+This is a magic built-in typeclass (similar to ``Coercible`` or
+``KnownSymbol``, for example).  It is given special treatment by the
+constraint solver, although users may also define their own instances,
+as described below.
+
+
+Solving HasField constraints
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+If the constraint solver encounters a constraint ``HasField x r a``
+where ``r`` is a concrete datatype with a field ``x`` in scope, it
+will automatically solve the constraint using the field selector as
+the dictionary, unifying ``a`` with the type of the field if
+necessary.  This happens irrespective of which extensions are enabled,
+as with other built-in classes with special constraint solving
+behaviour (e.g. ``Coercible``).
+
+For the ``personId`` example above, the end result is rather like
+having an instance
+
+.. code-block:: haskell
+
+  instance HasField "personId" Person Int where
+    fromLabel = personId
```

Comment:
Isn't the function inside `HasField` supposed to be `getField`? I thought `fromLabel` was from the `IsLabel` class... The other instance of `HasField` later in this document also defines `fromLabel` instead of `getField`.